File: Perceptron/Python/src/perceptron.py
import numpy as np 
from src.neuron import Neuron 


class Perceptron: 
    def __init__(self, num_inputs: int, lr: float):
        self.neuron = Neuron(num_inputs)
        self.lr = lr 


    def learning(self, x: np.ndarray, err: float, lr: float): # Perceptron's rule 
        w = self.neuron.getW()
        b = self.neuron.getB()
        
        w = w + lr * err * x  
        b = b + lr * err

        self.neuron.setW(w)
        self.neuron.setB(b)

    def predict(self, x: np.ndarray) -> int:
        u = self.neuron.soma(x)
        return self.neuron.activation(u)

    def train(self, X: np.ndarray, y: np.ndarray, num_epochs: int):
        for epoch in range(num_epochs):           
            for i in range(len(X)):
                y_pred = self.predict(X[i])
                err = y[i] - y_pred 

                if err != 0:
                    self.learning(X[i], err, self.lr)

File: Perceptron/C++/Perceptron.cpp
#include "headers/Perceptron.h"

using namespace std;

Perceptron::Perceptron(int num_inputs, double lr)
    : neuron(num_inputs), lr(lr) {}

Perceptron::~Perceptron() {}

void Perceptron::learning(const std::vector<double> &x, double err, double lr) {
  vector<double> w = this->neuron.getW();
  double b = this->neuron.getB();

  for (int i = 0; i < x.size(); ++i) {
    w[i] += lr * (err)*x[i];
  }
  b += lr * (err);

  this->neuron.setW(w);
  this->neuron.setB(b);
}

int Perceptron::predict(const vector<double> &x) {
  double u = this->neuron.soma(x);
  return this->neuron.activation(u);
}

void Perceptron::train(const vector<vector<double>> &X, const vector<int> &y,
                       int num_epochs) {

  for (int epoch = 0; epoch < num_epochs; epoch++) {
    for (int i = 0; i < X.size(); i++) {
      int error = y[i] - this->predict(X[i]);

      if (error != 0) {
        this->learning(X[i], error, this->lr);
      }
    }
  }
}
